daemon/graphdriver/copy/copy.go

```diff
@@ -15,26 +15,89 @@ import (
 	"golang.org/x/sys/unix"
 )
 
-type copyFlags int
+var copyFileRangeEnabled bool
+
+func init() {
+	copyFileRangeEnabled = isCopyFileRangeSyscallAvailable()
+}
+
+func isCopyFileRangeSyscallAvailable() bool {
+	if _, err := unix.CopyFileRange(-1, nil, -1, nil, 0, 0); err == unix.ENOSYS {
+		return false
+	}
+	return true
+}
+
+// The maximum amount of bytes Linux will splice
+const maxRwCount = 0x7ffff000
+
+// FICLONE ioctl number -- defined in <linux/fs.h> -- We don't include it
+// because it'd break builds with older versions of kernel headers
+const ficlone = 0x40049409
+
+// Mode indicates whether to use hardlink or copy content
+type Mode int
 
 const (
-	copyHardlink copyFlags = 1 << iota
+	// Content creates a new file, and copies the content of the file
+	Content Mode = iota
+	// Hardlink creates a new hardlink to the existing file
+	Hardlink
 )
 
-func copyRegular(srcPath, dstPath string, mode os.FileMode) error {
+// If useCopyWithFileRange is true, this will try to use the copy_file_range zero-copy syscall
+func copyRegular(srcPath, dstPath string, fileinfo os.FileInfo) error {
 	srcFile, err := os.Open(srcPath)
 	if err != nil {
 		return err
 	}
 	defer srcFile.Close()
 
-	dstFile, err := os.OpenFile(dstPath, os.O_WRONLY|os.O_CREATE, mode)
+	// If the destination file already exists, we shouldn't blow it away
+	dstFile, err := os.OpenFile(dstPath, os.O_WRONLY|os.O_CREATE|os.O_EXCL, fileinfo.Mode())
 	if err != nil {
 		return err
 	}
 	defer dstFile.Close()
 
-	_, err = pools.Copy(dstFile, srcFile)
+	_, _, err = unix.Syscall(unix.SYS_IOCTL, dstFile.Fd(), ficlone, srcFile.Fd())
```

Comment:
We don't need to attempt to do this clone first and should just rely on doing the copy file range syscall. My impression is the number of kernel/fileystem combos which will support this ioctl and do not have support for copy file range is small and do not justify making this attempt on every copy.

Reply:
You're right, every file system that supports FICLONE also supports copy_file_range by reflink.

1. The performance for large files of FICLONE seems to be better
2. clones are expected to be a fast metadata operation

What do you think about moving the FICLONE to only run iif every previous FICLONE was successful, and copyFileRangeEnabled is true?  There would be a nicety to this in that we could also keep track of exdev, and stop optimizing when we get our first exdev?

Reply:
> The performance for large files of FICLONE seems to be better

Are there benchmarks that show this? Which operations are done by copy file range which increase the overhead over clone. 

> What do you think about moving the FICLONE to only run iif every previous FICLONE was successful, and copyFileRangeEnabled is true?

I really think this is overcomplicating the code. Why not just attempt the copy file range call, and if it gets `== unix.ENOSYS` move on to legacy copy.  We should be doing the same thing we are doing in containerd here, see https://github.com/containerd/containerd/blob/master/fs/copy_linux.go#L34. Once we update to the latest version of containerd this code should merge with it. Based on the previous comment, the containerd should be updated to ensure files larger than 2GB are handled without error.

Reply:
Apparently Go handles this okay, even in a 32-bit environment:
```
package main

import (
    "fmt"
    "golang.org/x/sys/unix"
    "os"
)

func main() {
    f1, _ := os.Open("f1")
    f1Info, _ := os.Stat("f1")
    fmt.Println("F1 Size: ", f1Info.Size())

    f2, _ := os.OpenFile("f2", os.O_RDWR|os.O_TRUNC|os.O_CREATE, 0777)
    amountToCopy := f1Info.Size()
	fmt.Println("X: ", int(amountToCopy))
	fmt.Println(int(amountToCopy) == -1073741823)
    for amountToCopy > 0 {
        n, _ := unix.CopyFileRange(int(f1.Fd()), nil, int(f2.Fd()), nil, int(amountToCopy), 0)
        amountToCopy = amountToCopy - int64(n)
    }
}

```
Output:
```
root@titusagent-devvpc7-m4:/# strace -e trace=285 ./main 
F1 Size:  3221225473
X:  -1073741823
true
syscall_377(0x3, 0, 0x4, 0, 0xc0000001, 0) = 0x7ffff000
syscall_377(0x3, 0, 0x4, 0, 0x40001001, 0) = 0x40001001
+++ exited with 0 +++
```

Reply:
Current code looks good to me. 
@dmcgowan, there is no difference with relevant file systems between clone_file_range() and copy_file_range() implementation. The only difference is the MAX_RW_COUNT limit and return value of copied bytes from copy_file_range vs. return value 0 from clone_file_range(), hence the line ret = len; only after clone_file_range().
FYI, that copy_file_range() syscall tries to call clone_file_range() internally is a recent addition to kernel, far beyond v4.5 and the only way to grammatically test it is to try copy_file_range() with more than MAX_RW_COUNT bytes.

Reply:
@amir73il thanks for that explanation, I would not have been able to understand that background on my own. Sounds like you are advocating then for keeping the clone attempt because only the most recent kernel's `copy_file_range` call is as fast as the clone call, due to the recent update to have it `clone_file_range` internally? This sounds reasonable to me to keep as is in that case.

Reply:
I am saying current code is fine, because it won't try to clone more than once anyway. Its not that clone_file_range is that much faster than copy_file_range on xfs/btrfs. They are both the same fast metadata operations, only copy_file_range will limit metadata operation to 32bit range, so copying a very large file will require a few more syscalls. probably not a very common case for docker to deal with.